crimson/osd/osd_operations/client_request: make loading-obc concurrent

cyx1231st opened the pull request:
Major changes:
1. Introduce `std::optional<future<>> enter_stage_maybe_sync()` so that the ordering can be preserved upon entering an ordered concurrent phase.
2. Wrap `with_obc()` in ordered concurrent phases and use `enter_stage_maybe_sync()`, so that loading missing obcs from object store can become concurrent.

Unfortunately, I'm not able to see improved performance in the local tests. So I dig into the waiting-for-load-obc issue which was initially observed:

Firstly, the issue isn't related to the read path. The hit ratio of obc-lru cache is good after increasing size to 512 (https://github.com/ceph/ceph/pull/55188), so normally read won't load missing obcs frequently.

Secondly, the issue only appears with alienstore -- reading from seastore is so fast that loading missing obcs won't accumulate.

I'm able to reproduce the issue using rados bench random 4K write to stress a 32-core crimson-osd/alienstore, it seems rados bench will always write to new objects in this case, so that obc always needs to be loaded from the store.

In this scenario and in the worst case, I saw in one shard, (e.g.) 2093 client-requests were waiting outside the exclusive `recover_missing` phase, while 11 requests were pending inside `recover_missing`, and the rest 55 requests were pending inside the concurrent `wait_repop` phase. Normally however, the issue isn't that serious, (e.g.) only 4/4/49 requests were in the according phases.

Further analysis shows the requests inside `recover_missing` were mostly pending to load missing obcs from the alienstore. 

After this PR, in the worst case and in one shard, (e.g.) 8 requests were pending in the newly introduced concurrent `lock_obc` phase, and the rest 3108 requests were pending in the concurrent `wait_repop` phase. Normally, (e.g.) 15/48 requests were in the according phases. This means all requests were pending in the alienstore.

In short, after this PR, most of the IO requests are effectively pending in the store in the random 4K write rados-bench stressed scenario, and performance-not-improved is probably due to that the alienstore/bluestore was already saturated during my tests.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

Matan-B commented:
> I'm able to reproduce the issue using rados bench random 4K write to stress a 32-core crimson-osd/alienstore, it seems rados bench will always write to new objects in this case, so that obc always needs to be loaded from the store.

From previous discussions, Sam suggested to no longer use rados bench for perf testing exactly for this reason. IIUC, this is not an expected workload. @athanatos, please correct me if I'm wrong. 

---

`enter_stage_maybe_sync` is introduced solely for concurrent obc-loading which from the observation mentioned, async load is not a bottleneck in our current performance state.
What is the motivation behind introducing `enter_stage_maybe_sync` with it's (unavoidable) complexity burden if no performance improvement is obtained under expected workloads (assuming previous section is correct)?
IMO, we should trade simplicity over performance only if justified.
@cyx1231st, what do you think?

Note: The same applies with the way we handle MOSDMap, we have decided to avoid the complexity of handling concurrent MOSDMap messages in https://github.com/ceph/ceph/pull/51961 assuming the expected workloads shouldn't actually gain performance improvements from concurrent handling. The thrash tests would probably benefit from this but they only represent a testing workload.

athanatos commented:
Yep, rados bench simply creates a sequence of objects of the specified size -- that's it.  If you want to emulate a block workload, use fio.

cyx1231st commented:
> Note: The same applies with the way we handle MOSDMap, we have decided to avoid the complexity of handling concurrent MOSDMap messages in https://github.com/ceph/ceph/pull/51961 assuming the expected workloads shouldn't actually gain performance improvements from concurrent handling. The thrash tests would probably benefit from this but they only represent a testing workload.

I'm looking at this from a different angle: this change is directly related to the **critical IO path**, a more appropriate/generally-better design might still be worthwhile even if we cannot prove real benefits yet, specifically:
1. From Crimson-OSD perspective, it should be a generally better design to issue requests to the ObjectStore concurrently in the critical path.
2. The behavior of the ObjectStore is unpredictable -- there are different options of stores, their implementations are kept evolving, different hardware (CPU/disk) result in different performance patterns, the patterns from clients are various, the store behavior will vary due to other maybe unexpected factors (disk fullness, running time, background tasks, etc.).

Another challenge is that I can reliably reproduce the observation that "Critical IO requests can be stucked inside Crimson-OSD, waiting outside the OSD recover-missing phase because it is exclusive". **But I cannot conclude how much this will impact the end performance without applying this PR.** It might be none in this specific test case as the AlienStore seems to be already saturated, but it might be different in the future / with different hardware / with other unexpected factors (due to 2. above).

So from the OSD performance perspective, I think the evaluation complexity can be reduced with this PR (due to 1. and the challenge above).

cyx1231st commented:
> enter_stage_maybe_sync is introduced solely for concurrent obc-loading which from the observation mentioned, async load is not a bottleneck in our current performance state.
What is the motivation behind introducing enter_stage_maybe_sync with it's (unavoidable) complexity burden if no performance improvement is obtained under expected workloads (assuming previous section is correct)?
IMO, we should trade simplicity over performance only if justified.
@cyx1231st, what do you think?

I don't really agree that the "maybe_sync" change is more complex -- IIUC it is more general/explict to the fact that entering a new stage is not always meant to be asynchrous, by exposing `std::optional<future>` directly to users to respect this fact.

And it doesn't seem complex to me to transform `PipelineHandle::enter()` into a lightweight wrapper to the more general `PipelineHandle::enter_maybe_sync()` interface, so that the existing users don't need to be changed to adopt `std::optional<future>`.

cyx1231st commented:
> Yep, rados bench simply creates a sequence of objects of the specified size -- that's it. If you want to emulate a block workload, use fio.

I have to admit that we are still trying to understand how to use fio to properly stress the OSD -- we are still in this direction.

For example: If we want to compare the behavior of a stressed 1-core OSD and 32-core OSD, with the 1-core OSD got stressed by a 1-core fio client to 1 rbd image, which of the following ways is in-theory better (effective, fair and simple enough) to stress a 32-core OSD?
1. Use 1 fio client running with 32 jobs to 1 rbd image, within 32 cores;
2. Use 32 fio clients each with 1 job, all to to the same rbd image, within 32 cores;
3. Use 32 fio clients each with 1 job, each to an independent rbd image, within 32 cores;

For 3., I think we also need to adjust pg numbers and image sizes properly to make sure comparing a stressed 32-core OSD with a 1-core OSD baseline is fair.

Currently we use rados bench as an initial step to stress OSDs, so that we don't need (yet) to worry about the extra complexities from rbd, before we have enough experiments to be confident about the method.

cyx1231st commented:
> From previous discussions, Sam suggested to no longer use rados bench for perf testing exactly for this reason. IIUC, this is not an expected workload. @athanatos, please correct me if I'm wrong.

> Yep, rados bench simply creates a sequence of objects of the specified size -- that's it. If you want to emulate a block workload, use fio.

For this PR, specifically, I presumed that rados bench may be sufficient (from OSD perspective) to simulate/reproduce the case when the client IO pattern penetrates the obc-lru, so that obcs are mostly loaded from the store.

cyx1231st commented:
Just let you know that I'm not able to respond in time in the following 2 weeks.

liu-chunmei commented on src/crimson/osd/osd_operations/client_request.cc:
```diff
@@ -222,47 +222,45 @@ ClientRequest::process_op(
     client_pp(*pg).recover_missing, *this
-  ).then_interruptible([pg, this, FNAME]() mutable {
+  ).then_interruptible([pg, this, FNAME] {
     const hobject_t &soid = m->get_hobj();
```
soid seems not needed here.

cyx1231st replied:
fixed inplace.

liu-chunmei commented:
load obc will call backend store, so how fast depend on store response(just read attr may return fast depend on the DB size and search speed.). But seems no much work to do after the last exclusive phase and before this concurrent phase, and the same thread read SSD, seems can't take benefit from the parallel mechanism in SSD, that maybe you didn't see performance improvement.  The cleanup is reasonable, and if no performance degrade,  for store read use concurrent will not block the work before it.

Matan-B commented on src/crimson/common/operation.h:
```diff
@@ -506,6 +501,18 @@ class PipelineHandle {
+      return seastar::now();
+    });
+  }
+
```
:+1:

Matan-B commented on src/crimson/common/operation.h:
```diff
@@ -502,15 +502,22 @@ class PipelineHandle {
-    return std::move(fut).then(
-      [this, t=std::move(t)](auto &&barrier_ref) mutable {
+  std::optional<seastar::future<>>
+  do_enter_maybe_sync(T &stage, typename T::BlockingEvent::template Trigger<OpT>&& t) {
```
Can you please add a comment or a commit message here explanation here?

Matan-B replied:
nit: do_enter_maybe_block?

athanatos replied:
I can't quite follow what the intent here is.  This function needs a comment explaining what do_enter_maybe_sync does and what the return value means.

Matan-B commented on src/crimson/common/operation.h:
```diff
@@ -502,15 +502,22 @@ class PipelineHandle {
-      [this, t=std::move(t)](auto &&barrier_ref) mutable {
+  std::optional<seastar::future<>>
+  do_enter_maybe_sync(T &stage, typename T::BlockingEvent::template Trigger<OpT>&& t) {
+    if constexpr (T::is_enter_async) {
```
nit: is_blocking_enter

cyx1231st replied:
changed to `is_enter_sync` following the suggestion around introducing a safer version of `enter_stage_sync()`

Matan-B commented on src/crimson/osd/osd_operation_external_tracking.h:
```diff
@@ -103,6 +106,20 @@ struct LttngBackend
+              const ClientRequest::PGPipeline::RecoverMissingLockOBC& blocker) override {
+  }
+
+  void handle(ClientRequest::PGPipeline::RecoverMissingLockOBC::BlockingEvent::ExitBarrierEvent& ev,
```
line lengths around this commit

cyx1231st replied:
fixed inplace

Matan-B reviewed (approved):
Overall LGTM, some comments.
Naming changes are suggestions only, same as we have `maybe_record_blocking`. I think that "block" is friendlier than "sync".

Matan-B commented:
> I have to admit that we are still trying to understand how to use fio to properly stress the OSD -- we are still in this direction.
> 
> For example: If we want to compare the behavior of a stressed 1-core OSD and 32-core OSD, with the 1-core OSD got stressed by a 1-core fio client to 1 rbd image, which of the following ways is in-theory better (effective, fair and simple enough) to stress a 32-core OSD?
> 
> 1. Use 1 fio client running with 32 jobs to 1 rbd image, within 32 cores;
> 2. Use 32 fio clients each with 1 job, all to to the same rbd image, within 32 cores;
> 3. Use 32 fio clients each with 1 job, each to an independent rbd image (in different rados pools), within 32 cores;
> 
> For 3., I think we also need to adjust pg numbers and image sizes properly to make sure comparing a stressed 32-core OSD with a 1-core OSD baseline is fair.
> 
> Currently we use rados bench as an initial step to stress OSDs **by cores**, so that we don't need (yet) to worry about the extra complexities from rbd, before we have enough experiments to be confident about the appropriate method using fio.

What is the difference between running n clients vs 1 client running n jobs?

Regarding the separate image per client, I'm not sure we should go this route yet. For this stage, I think that single image is more suitable for our comparisons at the moment. (Will try to investigate)

(Let's move this discussion to the Slack channel since it's out of scope for this PR and will be easier to follow.)

athanatos commented on src/crimson/common/operation.h:
```diff
@@ -523,26 +537,32 @@ class PipelineHandle {
+    if (ret.has_value()) {
+      return std::move(ret.value());
+    } else {
+      return seastar::now();
```
Why not just return a ready future from enter_maybe_sync?

cyx1231st replied:
Upon leaving a concurrent phase and entering an exclusive phase, `enter_maybe_sync()` can return a non-ready future.

So its "maybe" sync.

athanatos replied:
I mean, if ret were a ```seastar::future<>``` rather than ```std::optional<seastar::future<>>``` and you unconditionally returned it, it seems like the behavior would be identical?

athanatos replied:
I suppose the recover_missing_snaps portion is the important one here.

cyx1231st replied:
I'm not sure if I get you question correctly.

For example, if ret were a `seastar::future<>` in the above example (leaving a concurrent phase and entering an exclusive phase), returning it to the caller should be correct to preserve ordering.

Wating for that future to be resolved means that the request has left the concurrent phase in order, and has already enterred the follow-up exclusive phase in the correct order.

cyx1231st replied:
> Why not just return a ready future from enter_maybe_sync?

Hm, probably I'm answering in the wrong direction.

`enter_maybe_sync()` is supposed to return non-future `nullopt` to support `lock_obc()` implementations.

There are many other existing users of `enter()` that don't expect `std::option<future>`, and I think it is unnecessary to adjust them, so `enter()` wraps `enter_maybe_sync()` to always return a future like before.

athanatos commented on src/crimson/osd/osd_operations/client_request.cc:
```diff
@@ -215,15 +215,79 @@ auto ClientRequest::reply_op_error(const Ref<PG>& pg, int err)
+          return recover_missing_snaps(pg, ihref, snaps);
+        });
+      } else {
+        return recover_missing_snaps(pg, ihref, snaps);
```
It seems like this std::optional<seastar::future> return from enter_stage_maybe_sync is to avoid a .then() on a ready future.  Is that actually worth the extra complexity?

cyx1231st replied:
> It seems like this std::optional<future> return from enter_stage_maybe_sync is to avoid a .then() on a ready future. Is that actually worth the extra complexity?

Please refer to a similar change https://github.com/ceph/ceph/commit/c7628de4d230cd08d536d7e0c35bf6822cc7146e

The goal is to avoid `.then()` upon leaving an exclusive and entering a concurrent phase, to preserve the correct calling order of `recover_missing_snaps()` or `process_lock_obc()`, specifically, the calling order of `pg->obc_loader.with_obc()` or `pg->with_locked_obc()`.

The obc lock might be exclusive for the same object, and once locked, it will be held across multiple phases, so I suppose if the locking order of the same object is reordered, we will encounter dead-lock.

cyx1231st replied:
Hm, probably I should instead check `ceph_assert(!maybe_fut.has_value())` here and with proper comment.

athanatos replied:
So the caller must ensure that with_obc() (through recover_missing_snaps) is called synchronously following the call to ihref.enter_stage_maybe_sync, but only if std::nullopt is returned?

cyx1231st replied:
I suppose in case of `with_obc()`, we should expect that the return value is always `std::nullopt` because `ihref.enter_stage_maybe_sync` is supposed to leave an exclusive phase (won't yield) and enter a concurrent phase (also won't yield) at the same time.

This is why I think the assert can be added https://github.com/ceph/ceph/pull/55488#discussion_r1506903525

athanatos replied:
For this call site, we are transitioning from recover_missing (OrderedExclusive) to recover_missing_lock_obc (OrderedConcurrent).  We can be certain that transition will occur without blocking because exiting an OrderedExclusivePhase never blocks and entering an OrderedConcurrent phase never blocks.  However, there isn't enough type information in PipelineHandle to prove the above statically -- PipelineHandle::barrier could be populated.

The next call site

```
           // call process_lock_obc() in order, but wait concurrently for loading.
            auto maybe_fut = ihref.enter_stage_maybe_sync(
                client_pp(*pg).lock_obc, *this);
```

is a transition from get_obc (OrderedExclusive) to lock_obc (OrderedConcurrent).  Same story -- we know that this call will always return std::nullopt, but the type information in PipelineHandle is insufficient to prove it.

The final call site is in internal_client_request:

```
            // call process_lock_obc() in order, but wait concurrently for loading.
            auto maybe_fut = enter_stage_maybe_sync(client_pp().lock_obc);
            if (maybe_fut.has_value()) {
              return interruptor::make_interruptible(
                std::move(maybe_fut.value())
              ).then_interruptible([this, &osd_ops] {
                return process_lock_obc(osd_ops);
              });
```

This one is again a transition from get_obc (OrderedExclusive) to lock_obc (OrderedConcurrent) and will never block.

It seems to me that:
1. In all three cases, as currently written, enter_stage_maybe_sync will *always* return std::nullopt
2. If 1. ever changes, it will introduce a subtle bug as the transition to the second stage won't be synchronous with invoking with_obc()

So, I suggest that we replace enter_stage_maybe_block with enter_stage_sync.  enter_stage_sync returns void, statically asserts !T::is_enter_async, and dynamically asserts that barrier->wait() returns std::nullopt.

I am still fuzzy on the details here, so I may well be missing something, though.

cyx1231st replied:
Yeah, reasonable to me.

cyx1231st replied:
see https://github.com/ceph/ceph/pull/55488/commits/61ddc7499aefcf47a3e5167290880be902d936e2

athanatos reviewed (changes requested):
I think I'm missing something with enter_stage_maybe_sync.

cyx1231st commented:
@athanatos reviews should be addressed, please take a look.

cyx1231st commented on src/crimson/osd/osd_operations/client_request.cc:
```diff
@@ -215,15 +215,71 @@ auto ClientRequest::reply_op_error(const Ref<PG>& pg, int err)
+    return do_recover_missing(pg, m->get_hobj().get_head());
+  }).then_interruptible([pg, this, &ihref] {
+    std::set<snapid_t> snaps = snaps_need_to_recover();
+    if (snaps.empty()) {
```
Seems CI is breaking due to a missing `!pg->is_primary()` check.

cyx1231st replied:
fixed inplace

cyx1231st commented:
Something seems missing:
`src/crimson/common/tri_mutex.cc:143: void tri_mutex::promote_from_write(): Assertion `writers == 1' failed.`

cyx1231st commented:
I think the issue is related to https://github.com/ceph/ceph/blob/622068ce1e8ec561adc28b84f7edaf20ac7179a5/src/crimson/osd/object_context.h#L124-L131

The lock should be promoted immediately upon locked -- this is why there are `assert(readers == 1);` `assert(writers == 1);` in the according promotion logic. However, `lock.lock().then(...)` above seem to be problematic because if another same lock takes the lock during `then()` before the promotion, it will cause readers/writers > 1.

I think the described issue was hidden because the logic was protected by an exclusive lock. And the issue appears because this PR changes it to become really concurrent.

cyx1231st commented:
> Something seems missing:
> src/crimson/common/tri_mutex.cc:143: void tri_mutex::promote_from_write(): Assertion writers == 1' failed.`

Should be fixed by https://github.com/ceph/ceph/pull/55488/commits/3a6332fd6676da590b9ede46954b2a6a74308bd7, no longer happened in the tests.

cyx1231st commented:
baseline: https://pulpito.ceph.com/yingxin-2024-04-02_02:43:05-crimson-rados-ci-yingxin-crimson-load-obc-3-base-distro-default-smithi/
* 7635282, 7635287, 7635293: https://tracker.ceph.com/issues/62162
* 7635302: https://tracker.ceph.com/issues/63647

cyx1231st commented:
test: https://pulpito.ceph.com/yingxin-2024-04-02_02:45:18-crimson-rados-ci-yingxin-crimson-load-obc-4-distro-default-smithi/
3 failures: 7635326,7635336,7635340

OSDs didn't break in the above cases. Not sure the root causes yet... From osd logs, seems "7635336" is related to scrubbing.

athanatos commented:
> test: https://pulpito.ceph.com/yingxin-2024-04-02_02:45:18-crimson-rados-ci-yingxin-crimson-load-obc-4-distro-default-smithi/ 3 failures: 7635326,7635336,7635340
> 
> OSDs didn't break in the above cases. Not sure the root causes yet... From osd logs, seems "7635336" is related to scrubbing.

Looks like 7635336 passed and 7635334 failed.

I took a look at 7635326 --

The actual failure was:

```
2024-04-02T03:32:24.321 INFO:tasks.rados.rados.0.smithi098.stderr:Error: finished tid 2 when last_acked_tid was 3
```

Looks like two writes on smithi09831302-718 reordered.

```
DEBUG 2024-04-02 03:32:24,087 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 571'2286 (0'0,571'2286] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 luod=571'2287 lua=0'0 crt=571'2287 mlcod 571'2286 active+clean  ClientRequest::process_op: client_request(id=20932, detail=m=[osd_op(client.4226.0:9780 3.1 3:8463c7bb:::smithi09831302-718:head {write 1756959~430975 in=430975b} snapc 12a={
12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: past scrub blocker, getting obc
...
DEBUG 2024-04-02 03:32:24,101 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 572'2288 (0'0,572'2288] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 luod=571'2287 lua=0'0 crt=572'2289 lcod 571'2286 mlcod 571'2286 active+clean  ClientRequest::process_op: client_request(id=20933, detail=m=[osd_op(client.4226.0:9781 3.1 3:8463c7bb:::smithi09831302-718:head {write 2781926~277410 in=277410b} snapc 12a={12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: past scrub blocker, getting obc
...
DEBUG 2024-04-02 03:32:24,192 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 572'2288 (0'0,572'2288] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 luod=572'2289 lua=0'0 crt=572'2289 mlcod 572'2288 active+clean  ClientRequest::process_lock_obc: client_request(id=20933, detail=m=[osd_op(client.4226.0:9781 3.1 3:8463c7bb:::smithi09831302-718:head {write 2781926~277410 in=277410b} snapc 12a={12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: got obc exists true oi 3:8463c7bb:::smithi09831302-718:head(572'2289 client.4226.0:9779  s 2050619 uv 2288 alloc_hint [0 0 0]), entering process stage
...
DEBUG 2024-04-02 03:32:24,193 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 572'2288 (0'0,572'2288] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 luod=572'2289 lua=0'0 crt=572'2289 mlcod 572'2288 active+clean  ClientRequest::process_lock_obc: client_request(id=20932, detail=m=[osd_op(client.4226.0:9780 3.1 3:8463c7bb:::smithi09831302-718:head {write 1756959~430975 in=430975b} snapc 12a={12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: got obc exists true oi 3:8463c7bb:::smithi09831302-718:head(572'2289 client.4226.0:9779  s 2050619 uv 2288 alloc_hint [0 0 0]), entering process stage
...
DEBUG 2024-04-02 03:32:24,194 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 572'2288 (0'0,572'2288] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 luod=572'2289 lua=0'0 crt=572'2289 mlcod 572'2288 active+clean  ClientRequest::process_lock_obc: client_request(id=20933, detail=m=[osd_op(client.4226.0:9781 3.1 3:8463c7bb:::smithi09831302-718:head {write 2781926~277410 in=277410b} snapc 12a={12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: in process stage, calling do_process
...
DEBUG 2024-04-02 03:32:24,202 [shard 1:main] osd -  pg_epoch 572 pg[3.1( v 572'2289 (0'0,572'2289] local-lis/les=14/15 n=89 ec=14/14 lis/c=14/14 les/c/f=15/15/0 sis=14) [3,1,0] r=0 lpr=14 lua=0'0 crt=572'2290 lcod 572'2288 mlcod 572'2288 active+clean  ClientRequest::process_lock_obc: client_request(id=20932, detail=m=[osd_op(client.4226.0:9780 3.1 3:8463c7bb:::smithi09831302-718:head {write 1756959~430975 in=430975b} snapc
 12a={12a, 129, 128, 127, 126, 123, 122, 121, 120, 11f, 11d, 11c, 11a, 119, 118, 117, 116, 114, 111, 10f, 10a, 109, 104, 102, 101, fe, fd, fc, fa, f9, f3, e4, de, cf, c8, c2, b5, b3, 76} ondisk+write+known_if_redirected+supports_pool_eio e572)]).0: in process stage, calling do_process
```

Simplified a bit:

```
...client.4226.0:9780...past scrub blocker, getting obc
...client.4226.0:9781...past scrub blocker, getting obc
...client.4226.0:9781...entering process stage
...client.4226.0:9780...entering process stage
...client.4226.0:9781...in process stage, calling do_process
...client.4226.0:9780...in process stage, calling do_process
```

Looks they reordered in the lock_obc stage.

I suggest we start by rebasing on main now that the coroutine changes to client_request merged -- that'll make it slightly easier to read.

cyx1231st commented:
Thanks for the help to parse the log!

> Looks like https://github.com/ceph/ceph/commit/763533612a10440951bfcddb2115c1f4bf8ca104 passed and 7635334 failed.

Sorry, pasted a wrong number.

> Looks they reordered in the lock_obc stage.

Yeah, somehow the phase doesn't work as expected to keep the order.

> I suggest we start by rebasing on main now that the coroutine changes to client_request merged -- that'll make it slightly easier to read.

Will rebase.

cyx1231st commented on src/crimson/common/operation.h:
```diff
@@ -523,29 +558,29 @@ class PipelineHandle {
-    if (wait_fut.has_value()) {
-      return wait_fut.value().then([this, &stage, t=std::move(t)] () mutable {
-        auto fut = t.maybe_record_blocking(stage.enter(t), stage);
-        exit();
```
@athanatos  I probably understand the issue here -- `exit()` should be moved after `fut` is resolved.

Specifically, the shared-mutex from concurrent `get-obc` phase should be released (by `exit()`) **after** the request has entered the next exclusive `process` phase.

athanatos replied:
Ah, that makes sense.  Approved the PR.

cyx1231st commented:
> I suggest we start by rebasing on main now that the coroutine changes to client_request merged -- that'll make it slightly easier to read.

Rebased and reworked this PR. @athanatos @Matan-B 

Depends on the fix https://github.com/ceph/ceph/pull/56844

cyx1231st commented:
Looks the reordering issue is still present and same.

https://pulpito.ceph.com/yingxin-2024-04-24_01:32:34-crimson-rados-ci-yingxin-crimson-load-obc-new-1-distro-default-smithi/7671131/  osd.2
```
...
DEBUG 2024-04-24 01:55:03,967 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'678 (0'0,227'678] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 luod=226'678 lua=0'0 crt=227'679 lcod 226'677 mlcod 226'677 active+clean  ClientRequest::process_op: client_request(id=16780312, detail=m=[osd_op(client.4240.0:2567 3.e 3:7f533780:::smithi04928907-48:head {write 2928991~447707 in=447707b} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: past scrub blocker, getting obc
...
DEBUG 2024-04-24 01:55:03,969 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'678 (0'0,227'678] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 luod=226'678 lua=0'0 crt=227'679 lcod 226'677 mlcod 226'677 active+clean  ClientRequest::process_op: client_request(id=16780313, detail=m=[osd_op(client.4240.0:2568 3.e 3:7f533780:::smithi04928907-48:head {setxattr _header (58) in=65b, truncate 3647185} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: past scrub blocker, getting obc
...
DEBUG 2024-04-24 01:55:03,984 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'678 (0'0,227'678] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 luod=227'679 lua=0'0 crt=227'679 mlcod 227'678 active+clean  ClientRequest::process_op: client_request(id=16780313, detail=m=[osd_op(client.4240.0:2568 3.e 3:7f533780:::smithi04928907-48:head {setxattr _header (58) in=65b, truncate 3647185} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: got obc exists true oi 3:7f533780:::smithi04928907-48:head(227'679 client.4240.0:2565  s 1150307 uv 678 alloc_hint [0 0 0]), entering process stage
...
DEBUG 2024-04-24 01:55:03,984 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'678 (0'0,227'678] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 luod=227'679 lua=0'0 crt=227'679 mlcod 227'678 active+clean  ClientRequest::process_op: client_request(id=16780312, detail=m=[osd_op(client.4240.0:2567 3.e 3:7f533780:::smithi04928907-48:head {write 2928991~447707 in=447707b} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: got obc exists true oi 3:7f533780:::smithi04928907-48:head(227'679 client.4240.0:2565  s 1150307 uv 678 alloc_hint [0 0 0]), entering process stage
...
DEBUG 2024-04-24 01:55:03,984 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'678 (0'0,227'678] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 luod=227'679 lua=0'0 crt=227'679 mlcod 227'678 active+clean  ClientRequest::process_op: client_request(id=16780313, detail=m=[osd_op(client.4240.0:2568 3.e 3:7f533780:::smithi04928907-48:head {setxattr _header (58) in=65b, truncate 3647185} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: in process stage, calling do_process
...
DEBUG 2024-04-24 01:55:03,989 [shard 1:main] osd -  pg_epoch 227 pg[3.e( v 227'679 (0'0,227'679] local-lis/les=15/16 n=7 ec=15/15 lis/c=15/15 les/c/f=16/16/0 sis=15) [2,1,3] r=0 lpr=15 lua=0'0 crt=227'680 lcod 227'678 mlcod 227'678 active+clean  ClientRequest::process_op: client_request(id=16780312, detail=m=[osd_op(client.4240.0:2567 3.e 3:7f533780:::smithi04928907-48:head {write 2928991~447707 in=447707b} snapc 0={} ondisk+write+known_if_redirected+supports_pool_eio e227)]).0: in process stage, calling do_process
...
```
Shorter version:
```
pg[3.e] client_request(id=16780312) past scrub blocker, getting obc
pg[3.e] client_request(id=16780313) past scrub blocker, getting obc
pg[3.e] client_request(id=16780313) entering process stage
pg[3.e] client_request(id=16780312) entering process stage
pg[3.e] client_request(id=16780313) in process stage, calling do_process
pg[3.e] client_request(id=16780312) in process stage, calling do_process
```

cyx1231st commented:
See:
* expermental logs: https://github.com/ceph/ceph-ci/tree/ci-yingxin-crimson-load-obc-new-experiment-1
* test result: https://pulpito.ceph.com/yingxin-2024-04-29_01:06:33-crimson-rados-ci-yingxin-crimson-load-obc-new-experiment-1-distro-default-smithi/7678220/
* In osd 0, the requests `client.4235.0:5184(id=16781865)` and `client.4235.0:5186(id=16781867)` are reordered in `process` stage like above.

And the log shows:
```
...
INFO  2024-04-29 01:29:40,138 [shard 2:main] osd - Concurrent exit barrier 16781865: locking
...
INFO  2024-04-29 01:29:40,144 [shard 2:main] osd - Concurrent exit barrier 16781867: locking
...
INFO  2024-04-29 01:29:40,208 [shard 2:main] osd - concurrent exit barrier 16781867: wait...
...
INFO  2024-04-29 01:29:40,208 [shard 2:main] osd - concurrent exit barrier 16781865: wait...
...
! INFO  2024-04-29 01:29:40,208 [shard 2:main] osd - do_enter_maybe_sync 16781867: async enter...
...
! INFO  2024-04-29 01:29:40,208 [shard 2:main] osd - do_enter_maybe_sync 16781865: async enter...
...
! INFO  2024-04-29 01:29:40,209 [shard 2:main] osd - concurrent exit barrier 16781867: sync unlock
...
! INFO  2024-04-29 01:29:40,214 [shard 2:main] osd - concurrent exit barrier 16781865: sync unlock
...
```

This suggests that 16781865 firstly locking `OrderedConcurrentPhaseT::mutex` then 16781867, but got locked in a reversed order, and thus unlocked in a wrong order.

However, looking at `seastar::shared_mutex`, the locking order should be guaranteed because `shared_mutex::_waiters` is an fifo queue.

cyx1231st commented:
Further investigations [1][2] showed that the issue is related to inappropriate exit (i.e. unlock) during request interruption, from a `OrderedExclusivePhaseT::mutex` (with the address `0x625000d81a08`) perspective:

```
...
465025:INFO  2024-05-06 01:47:47,871 [shard 1:main] osd - concurrent exit barrier: id=16778802 locking mutex=0x625000d81a08
* 465047:INFO  2024-05-06 01:47:47,873 [shard 1:main] osd - concurrent exit barrier: id=16778802 locked mutex=0x625000d81a08
465099:INFO  2024-05-06 01:47:47,874 [shard 1:main] osd - concurrent exit barrier: id=16778803 locking mutex=0x625000d81a08
465120:INFO  2024-05-06 01:47:47,875 [shard 1:main] osd - concurrent exit barrier: id=16778804 locking mutex=0x625000d81a08
465143:INFO  2024-05-06 01:47:47,877 [shard 1:main] osd - concurrent exit barrier: id=16778805 locking mutex=0x625000d81a08
! 465338:INFO  2024-05-06 01:47:47,882 [shard 1:main] osd - concurrent exit barrier: id=16778814 locking mutex=0x625000d81a08
! 465374:INFO  2024-05-06 01:47:47,883 [shard 1:main] osd - concurrent exit barrier: id=16778814 wait mutex=0x625000d81a08...
! 465377:INFO  2024-05-06 01:47:47,883 [shard 1:main] osd - concurrent exit barrier: id=16778814 sync unlock mutex=0x625000d81a08
* 465380:INFO  2024-05-06 01:47:47,883 [shard 1:main] osd - concurrent exit barrier: id=16778803 locked mutex=0x625000d81a08
465716:INFO  2024-05-06 01:47:47,888 [shard 1:main] osd - concurrent exit barrier: id=16778816 locking mutex=0x625000d81a08
465758:INFO  2024-05-06 01:47:47,889 [shard 1:main] osd - concurrent exit barrier: id=16778816 wait mutex=0x625000d81a08...
465760:INFO  2024-05-06 01:47:47,889 [shard 1:main] osd - concurrent exit barrier: id=16778816 sync unlock mutex=0x625000d81a08
465762:INFO  2024-05-06 01:47:47,889 [shard 1:main] osd - concurrent exit barrier: id=16778804 locked mutex=0x625000d81a08
465880:INFO  2024-05-06 01:47:47,891 [shard 1:main] osd - concurrent exit barrier: id=16778818 locking mutex=0x625000d81a08
465921:INFO  2024-05-06 01:47:47,892 [shard 1:main] osd - concurrent exit barrier: id=16778818 wait mutex=0x625000d81a08...
465925:INFO  2024-05-06 01:47:47,892 [shard 1:main] osd - concurrent exit barrier: id=16778818 sync unlock mutex=0x625000d81a08
465928:INFO  2024-05-06 01:47:47,892 [shard 1:main] osd - concurrent exit barrier: id=16778805 locked mutex=0x625000d81a08
465929:INFO  2024-05-06 01:47:47,892 [shard 1:main] osd - concurrent exit barrier: id=16778805 locked mutex=0x625000d81a08
466563:INFO  2024-05-06 01:47:47,947 [shard 1:main] osd - concurrent exit barrier: id=16778803 wait mutex=0x625000d81a08...
466588:INFO  2024-05-06 01:47:47,947 [shard 1:main] osd - concurrent exit barrier: id=16778804 wait mutex=0x625000d81a08...
466602:INFO  2024-05-06 01:47:47,947 [shard 1:main] osd - concurrent exit barrier: id=16778802 wait mutex=0x625000d81a08...
* 466641:INFO  2024-05-06 01:47:47,948 [shard 1:main] osd - concurrent exit barrier: id=16778803 sync unlock mutex=0x625000d81a08
! 466673:INFO  2024-05-06 01:47:47,948 [shard 1:main] osd - concurrent exit barrier: id=16778814 locked mutex=0x625000d81a08
* 466844:INFO  2024-05-06 01:47:47,952 [shard 1:main] osd - concurrent exit barrier: id=16778802 sync unlock mutex=0x625000d81a08
466874:INFO  2024-05-06 01:47:47,952 [shard 1:main] osd - concurrent exit barrier: id=16778816 locked mutex=0x625000d81a08
467032:INFO  2024-05-06 01:47:47,957 [shard 1:main] osd - concurrent exit barrier: id=16778804 sync unlock mutex=0x625000d81a08
...
```

Specifically, the op (id=16778814) incorrectly called unlock before it really held the lock (see !), which caused the op 16778803 to be locked unexpectedly. This is why it looked as if both op 16778802 and op 16778803 were holding the lock at the same time (see *)  , and finally led to incorrect order.

@athanatos @Matan-B 

[1] https://github.com/ceph/ceph-ci/tree/ci-yingxin-crimson-load-obc-new-experiment-2
[2] https://pulpito.ceph.com/yingxin-2024-05-06_01:28:05-crimson-rados-ci-yingxin-crimson-load-obc-new-experiment-2-distro-default-smithi/7693540/

cyx1231st commented:
Looks resolved: https://pulpito.ceph.com/yingxin-2024-05-08_01:32:38-crimson-rados-ci-yingxin-crimson-load-obc-new-2-distro-default-smithi/

Doing broader tests.

cyx1231st commented:
jenkins test make check

cyx1231st commented:
Changeset:
* Reworked the PR based on the coroutine changes
* Rebased to exclude the already merged commits
* Cleanup: https://github.com/ceph/ceph/pull/55488/commits/df2804dc1d4d9c983db407e145d69dcb1b065d3b
* Fix the reordering issue: https://github.com/ceph/ceph/pull/55488/commits/4a9ed13cb73361f4069de602c225e23002209203, looks it has resolved the issue

cyx1231st commented:
See results https://pulpito.ceph.com/yingxin-2024-05-08_02:24:47-crimson-rados-ci-yingxin-crimson-load-obc-new-2-distro-default-smithi/:
* 7697186: https://tracker.ceph.com/issues/64206
* 7697193: https://tracker.ceph.com/issues/64206
* 7697196: https://tracker.ceph.com/issues/65451

These issues exist before this PR. Allowing to load obc concurrently seem to intensify them (or make them easier to be reproduced in thrash tests).

@athanatos @Matan-B Are the test results look good to you?
I'm converting this PR from draft to open. I think it's ready to be reviewed again.
